### PR TITLE
add maps api to website

### PR DESCRIPTION
**Pull request by ntarn**
add markers for locations I am based at, fun.html can be updated later

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/BlobServingServlet.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import java.io.IOException;
+
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.annotation.WebServlet;
+
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/BlobServingServlet.java:11]: Alphabetical order.

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Servlet that creates a URL in the response to allow a user to upload a file
+ * to Blobstore when the fetch() function requests the /blobstore-upload-url
+ * URL. Navigate to /blobstore-upload-url to see the Blobstore URL.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java:28]: The line wrapping seems a bit off.

Great job on the comment though!

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.Query;
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:27]: The simple `.Query` would go first.

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+import com.google.gson.Gson;
+import com.google.sps.data.Comment;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Collections;
+import java.util.Map;
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:40]: Alphabetical order.

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Servlet that processes a form with a comment request. When the user submits
+ * the form, Blobstore processes the file upload and then forwards the request
+ * to this servlet. This servlet can then process the request using the file URL
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:49]: Nit:
```suggestion
 * to this servlet. The servlet can then process the request using the file URL
```

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+ * we get from Blobstore.
+ */
+@WebServlet("/my-form-handler")
+public class FormHandlerServlet extends HttpServlet {
+
+  private int previousMax = 1;
+  private final String COMMENT = "comment";
+  private final String TEXT = "text";
+  private final String TIMESTAMP = "timestamp";
+  private final String IMAGEURL = "imageUrl";
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:59]: How about making them `static` too

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the maximum amount of comments to display data from the server.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:63]:
```suggestion
    // Get the maximum number of comments to display from the server.
```

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    int maxCommentsObtained = getMaxComments(request);
+    if (previousMax == 1 && maxCommentsObtained != -1) {
+      previousMax = maxCommentsObtained;
+      System.err.println("ntarn debug previous max change to current max: " + maxCommentsObtained);
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:67]: Remove debug code.

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    } else if (maxCommentsObtained == -1) {
+      maxCommentsObtained = previousMax;
+      System.err.println("ntarn debug previous max: " + maxCommentsObtained);
+    } else {
+      System.err.println("ntarn debug previous max reach else statement: " + maxCommentsObtained);
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:72]: Ditto here too.

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    // Prepare a Query instance with the Comment kind of entity to load.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:75]: Could you rephrase this a little? It gets a bit confusing.

How about "Prepare a Query instance to load a Comment entity"

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    Query query = new Query(COMMENT).addSort(TIMESTAMP, SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:79]: You could combine these into one since you're not using `datastore` anywhere else.

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+    // Get the maximum number of comments that can be displayed on a page.
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();
+      String text = (String) entity.getProperty(TEXT);
+      long timestamp = (long) entity.getProperty(TIMESTAMP);
+      String imageUrl = (String) entity.getProperty(IMAGEURL);
+      Comment comment = new Comment(id, text, timestamp, imageUrl);
+      comments.add(comment);
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:89]: Combine these too by just inlining the `new Comment...` part.

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+   * didn't upload a file.
+   */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+    System.out.println("ntarn debug: blobKeys obtained");
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev
+    // server)
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:136]: I think the `server)` would fit on the previous line.

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live
+    // server)
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:145]: Here too?

File: portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    } else {
+      return blobKey.getKeyString();
+    }
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:152]: Use a conditional operator for the `return`

(Like we discussed a couple weeks ago over VC while debugging)

File: portfolio/src/main/webapp/js/customscript.js [hunk 16]
```diff
@@ -14,23 +14,25 @@
...
-      console.log(comments);
-      const commentListElement = document.getElementById('comment-list'); // Retrieve the list of comments at the ElementById.
-      commentListElement.innerHTML = "";
-      comments.forEach((comment) => {
-        commentListElement.appendChild(createCommentElement(comment));
-      })
+  const maxComments = document.getElementById('max-comments').value;
+  fetch('/my-form-handler?max-comments='+ maxComments)  // Sends a request to the URL.
+  .then(response => response.json()) // Parses the response as JSON.
+  .then((comments) => { // Now we can access the comments with this variable.
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/js/customscript.js:23]: Keep a consistent style.

```suggestion
  .then((comments) => { // Process comments.
```

File: portfolio/src/main/webapp/js/customscript.js [hunk 17]
```diff
@@ -40,6 +42,17 @@ function createCommentElement(comment) {
...
+  const imageUrlElement = document.createElement('img');
+  fetch(request).then(response => response.blob()).then((blob) => {
+    console.log('Adding images to dom: ' + comment.imageUrl);
+    imageUrlElement.src = window.URL.createObjectURL(blob);
+  })
+  
+  // const imageUrlElement = document.createElement('img');
+  // console.log('Adding images to dom: ' + comment.imageUrl);
+  // imageUrlElement.src = comment.imageUrl; 
+
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/js/customscript.js:55]: Remove unused code.

File: portfolio/src/main/webapp/js/mapscript.js [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+  var map = new google.maps.Map(document.getElementById('map'), {
+    center: sunnyvale,
+    zoom: 3,
+    mapTypeControlOptions: {
+      mapTypeIds: ['roadmap', 'satellite', 'hybrid', 'terrain',
+              'styled_map']
+    }
+  });
+
+  //Associate the styled map with the MapTypeId and set it to display.
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/js/mapscript.js:203]:
```suggestion
  // Associate the styled map with the MapTypeId and set it to display.
```

File: portfolio/src/main/webapp/js/mapscript.js [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+    '<div id="bodyContent">'+
+    '<p><b>Major: Computer Science and Biology</b><br>' +
+    'I aim to use data science to help with wildlife conservation and environmental issues. '+
+    'Currently, I am working on a natural language processing project that analyzes newspaper topics for the Duke Chronicle. '+
+    'Additionally, I am conducting research on identifying energy infrastructures using computer vision with Kyle Bradbury.'+
+    '</p>'+
+    '</div>'+
+    '</div>';
+  addLandmark(map, 37.402601, -122.015315, 'Sunnyvale', 'Where I\'m based');
+  addLandmark(map, 36.001339, -78.938218, 'Duke University', contentString);
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/js/mapscript.js:219]: Create constants for these points at the top of the file.

File: portfolio/src/main/webapp/js/mapscript.js [hunk 20]
```diff
@@ -0,0 +1,234 @@
+function createMap() {
+  // Create a new StyledMapType object, passing it an array of styles, and the name to be displayed on the map type control.
+  var styledMapType = new google.maps.StyledMapType(
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/js/mapscript.js:3]: Can we make this a const?

File: portfolio/src/main/webapp/js/mapscript.js [hunk 21]
```diff
@@ -0,0 +1,234 @@
+function createMap() {
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/js/mapscript.js:1]: I would add some details here in a comment.

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java (hunk 2 above)
Comment (ntarn) [reply portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java:28]: For the extremely long comments like these description ones, I did it based on a previous 80 character rule I learned before for comments, because when I checked the Google java guidelines, they said it based on the author's discretion, but let me know if there's another way you want me to do it.

Comment (hayleejane3) [reply portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java:28]: If this is 80 a line then its fine.
